### PR TITLE
Better reshape with autograd support

### PR DESCRIPTION
**Pull request by YifanShenSZ**
Previous reshape [#80981](https://github.com/pytorch/pytorch/pull/80981) is ok for forward, but needs improvement for backward: need to handle "sometimes view sometimes copy" behavior.

This pull request fixes it by:
1. add a new alias dispatch key `CompositeImplicitAutogradNestedTensor`, which ideally would work as nested-tensor version of `CompositeImplicitAutograd`
2. register `reshape_nested` to `reshape` by `CompositeImplicitAutogradNestedTensor`

Side changes:
* add contiguous memory format support to `clone_nested`
* add `view_nested`
* add `reshape_as_nested`

Fix issue [#83041](https://github.com/pytorch/pytorch/issues/83041)

### CODE REVIEW

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:708]: I think `empty_like` might be more clear here.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:705]: I would maybe add a comment that is the case when the memory format is contiguous but self is a non_contiguous nested tensor. 

One thing I am trying to hash out now. I think all current nested tensor instances satisfy the bottom two are equal.

`nested_tensor.numel()` -> this is calculated from nested_sizes to have less elements than its underlying 
storage ->` nested_tensor.storage.nbytes()//dtype`. 

In theory one can easily create this tensor with wrap_buffer(), however I am not sure if that is done anywhere. 

reshape, transpose, view / all current view ops appear to maintain this invariant but need to confirm.
@jbschlosser

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:1143]: if view_nested is returning a view it can't use get the storage of the self tensor and then returned a tensor that shares the underlying storage.

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp
Comment (albanD) [inline aten/src/ATen/core/dispatch/OperatorEntry.cpp:307]: You most likely want to update the comment above.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:705]: slice is an example to have `nested_tensor.numel()` < `nested_tensor.storage.nbytes()//dtype`
```
nt = torch.nested_tensor([x, y, z])
slice = nt[0 : 2 : 2]
```
where the memory of `y` is skipped

**Review by albanD — COMMENTED** [review]
separating adding the new alias key would allow you to land it without waiting for Driss's change.
I am not sure why view calls into reshape but didn't dive into this in details yet, maybe it is just a function name issue.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:1143]: Will update it according to your pr [#82658](https://github.com/pytorch/pytorch/pull/82658)

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:708]: will update it once nested `empty_like` comes out

Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:711]: That too but also meant you could just do `Tensor output_buffer = self_buffer.empty_like()`

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:694]: you should be able to call self.is_contiguous()

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:705]: This is essentially the impl of self.contiguous(), might be cleaner to just override contiguous() for nt and calling  self.contiguous() here, I can just note that for clean up though

Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:705]: I found `contiguous` a CompositeImplicit op, which may redispatch to `clone(Contiguous)`, so it seems to me that we should implement `clone(Contiguous)` then leave `contiguous` the same.

Anything I missed?

File: c10/core/DispatchKeySet.cpp
Comment (drisspg) [inline c10/core/DispatchKeySet.cpp:49]: Did this remove the ability for nested tensors to run in "regular" Composite kernels. We should have both right? And just have CompositeImplicitAutogradNestedTensor take precedence. @bdhirsh

File: torchgen/model.py
Comment (drisspg) [inline torchgen/model.py:663]: could be `dispatch_key in {DispatchKey.CompositeImplicitAutograd, DispatchKey.CompositeImplicitAutogradNestedTensor}`

File: c10/core/DispatchKeySet.cpp
Comment (YifanShenSZ) [reply c10/core/DispatchKeySet.cpp:49]: I think yes, but I have modified codegen in such a way that CompositeImplicitAutogradNetsedTensor kernel got registered even if only CompositeImplicitAutograd is specified. Take `softmax` as an example, in native_functions.yaml it has:
```
- func: softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
  variants: function, method
```
which makes it CompositeImplicitAutograd. A `torch._C._dispatch_dump("aten::softmax.int")` call gives us:
```
name: aten::softmax.int
schema: aten::softmax.int(Tensor self, int dim, int? dtype=None) -> Tensor
debug: registered at aten/src/ATen/RegisterSchema.cpp:6
alias analysis kind: FROM_SCHEMA
Functionalize: registered at aten/src/ATen/RegisterFunctionalization_2.cpp:20485 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
Named: registered at ../aten/src/ATen/core/NamedRegistrations.cpp:11 :: (none) [ fallthrough boxed ]
Tracer: registered at ../torch/csrc/autograd/generated/TraceType_3.cpp:14597 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
AutocastCUDA: registered at ../aten/src/ATen/autocast_mode.cpp:328 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
CompositeImplicitAutograd[alias]: registered at aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6983 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
CompositeImplicitAutogradNestedTensor[alias]: registered at aten/src/ATen/RegisterCompositeImplicitAutogradNestedTensor.cpp:6922 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
```
Which has CompositeImplicitAutogradNestedTensor registered as well

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:705]: ohh cool thats perfect. So this change should allow us to call `.contiguous` on nested tensors? Maybe we should add a test for this

File: c10/core/DispatchKeySet.cpp
Comment (drisspg) [reply c10/core/DispatchKeySet.cpp:49]: Got it so the registered kernel is the same for CompositeImplicitAutograd , CompositeImplicitAutogradNestedTensor when CompositeImplicitAutogradNestedTensor is not specified? This makes sense

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (YifanShenSZ) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:705]: Just tried. Yes we can call `nested_tensor.contiguous()` now

File: aten/src/ATen/native/native_functions.yaml
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:6931]: Doesn't need to this PR, but we should consider renaming the existing `DispatchKey::NestedTensor` to `DispatchKey::NestedTensorFunctionality`, and then making `DispatchKey::NestedTensor` a new alias key for "all nested tensor backends". That way you don't have special-case for cpu/cuda and in the other places.

File: c10/core/DispatchKey.h
Comment (bdhirsh) [inline c10/core/DispatchKey.h:443]: hmmm I think you'll need to make internal fbcode changes, since this adds a new codegen'd file.

I would fbgs search for `RegisterCompositeImplicitAutograd.cpp` to see where that file gets hardcoded into the build system, to see where you need to add it

File: c10/core/DispatchKeySet.cpp
Comment (bdhirsh) [reply c10/core/DispatchKeySet.cpp:49]: hmm I'm not sure that this is what we want.

I think of `CompositeImplicitAutograd` as "all backends + layout keys, plus all autograd keys".

And then now, I think of `CompositeImplicitAutogradNestedTensor` as "all nested backends, plus the nested autograd key".

So in that view, `CompositeImplicitAutograd` should be a strict superset of `CompositeImplicitAutogradNestedTensor`: if a kernel is registered directly to `CompositeImplicitAutograd`, that should still cause all of the nested tensor dispatch keys to get that kernel too.

We also have a handful of native_function entries that explicit use `CompositeImplicitAutograd`, so this would cause NestedTensor to lose those kernels. e.g. https://github.com/pytorch/pytorch/blob/1f38225b56b873c944196241ea448445a61798fd/aten/src/ATen/native/native_functions.yaml#L3951.

File: tools/autograd/derivatives.yaml
Comment (bdhirsh) [inline tools/autograd/derivatives.yaml:1730]: This is cool :)

File: torchgen/gen.py
Comment (bdhirsh) [inline torchgen/gen.py:1250]: This looks like it'll change the output of "dispatch" for existing native functions that only have a `CompositeImplicitAutograd` entry.

Feels kinda ugly, but I guess you want to check for `... != {CompositeImplicitAutograd} and ... != {CompositeImplicitAutograd, CompositeImplicitAUtogradNestedTensor}`

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:664]: I think we can kill this if you follow feedback from https://github.com/pytorch/pytorch/pull/82754/files#r950515109

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:994]: Is it the case that "if an op has a CompositeImplicitAutogradNestedTensor kernel, then it always has a CompositeImplicitAutograd kernel"?

If it is, then we can probably take advantage of that to simplify this: just assert that in `model.py`, and you don't need this extra check.

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:2563]: this check makes me sad (because its confusing) but I guess we technically need to guard against it :p

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:2582]: same comment as https://github.com/pytorch/pytorch/pull/82754/files#r950522386 (I don't think we need this)

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:2594]: same as https://github.com/pytorch/pytorch/pull/82754/files#r950522386

File: c10/core/DispatchKeySet.cpp
Comment (YifanShenSZ) [reply c10/core/DispatchKeySet.cpp:49]: Have added NestedTensor back to math keyset

File: c10/core/DispatchKey.h
Comment (YifanShenSZ) [reply c10/core/DispatchKey.h:443]: Have to ask you guys to do it if it turns out necessary :stuck_out_tongue_winking_eye:

File: tools/autograd/derivatives.yaml
Comment (YifanShenSZ) [reply tools/autograd/derivatives.yaml:1730]: credit to Mikayla

File: torchgen/model.py
Comment (YifanShenSZ) [reply torchgen/model.py:994]: I guess usually it is, but there may be ops designated to nested tensor (e.g. `to_padded_tensor`)

@drisspg @jbschlosser what do you guys think?

Comment (drisspg) [reply torchgen/model.py:994]: For all our current ops I think that is the case but I am not 100% that is the case forever

Comment (drisspg) [reply torchgen/model.py:993]: we do need this I wouldn't change

File: test/test_nestedtensor.py
Comment (bdhirsh) [inline test/test_nestedtensor.py:238]: huh, this PR fixed `.contiguous()` for nested tensors?

Comment (bdhirsh) [reply test/test_nestedtensor.py:238]: oh nevermind - just re-read the PR description :)

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:703]: I think we still want to get rid of this, as per https://github.com/pytorch/pytorch/pull/82754#discussion_r950515109.

In the yaml, this:
```
func: foo(...) -> ...:
```
is syntactic sugar for:
```
func foo(...) -> ...:
  dispatch:
    CompositeImplicitAutograd: foo
```

That shouldn't need any extra changes for nested tensors, because nested tensors always opt into `CompositeImplicitAutograd` (unless they have a kernel registered to `CompositeImplicitAutogradNestedTensor`, which takes higher priority)

File: torchgen/model.py
Comment (bdhirsh) [inline torchgen/model.py:911]: Is this function actually used anywhere in the codegen?

I just grepped around and I don't think it is. If that's the case, do you mind deleting it in this PR?

It would be a pretty confusing function to use now that we have `CompositeImplicitAutogradNestedTensor`: you might have a function that returns `f.has_composite_kernel() == True`, but it only has a decomposition that works for nested tensors, and not for normal tensors, which probably isn't what you expect.

File: torchgen/model.py
Comment (bdhirsh) [reply torchgen/model.py:993]: Agreed, now that we have https://github.com/pytorch/pytorch/pull/83371/files#r950576993

File: torchgen/native_function_generation.py
Comment (bdhirsh) [inline torchgen/native_function_generation.py:393]: I think this could actually break stuff - we should just remove this line. (too bad nothing failed in CI to tell you that, probably because it didnt' affect any existing operators today).

The idea here is that all operators in native_functions.yaml (that do **not** have `CompositeImplicitAutograd` kernels) are now required to have functional + out= variants.

It's now technically possible to have an op with a `CompositeImplicitAutogradNestedTensor` kernel, but not a `CompositeImplicitAutograd` kernel. We would still want the out= and functional variants to be generated for that operator.

**Comment by bdhirsh** [comment]
Left some final comments on the codegen but otherwise I think this PR looks good!

File: torchgen/model.py
Comment (YifanShenSZ) [reply torchgen/model.py:777]: changed to
```
is_abstract = dispatch.keys() != {DispatchKey.CompositeImplicitAutograd} \
            and dispatch.keys() != {DispatchKey.CompositeImplicitAutogradNestedTensor} \
            and dispatch.keys() != {
                DispatchKey.CompositeImplicitAutograd,
                DispatchKey.CompositeImplicitAutogradNestedTensor,
            }
```

File: torchgen/model.py
Comment (YifanShenSZ) [reply torchgen/model.py:911]: Tried deleting it but got errors. It should be used in some way.

Changed to
```
    @property
    def has_composite_kernel(self) -> bool:
        return (
            self.has_composite_implicit_autograd_kernel
            or self.has_composite_explicit_autograd_kernel
            or self.has_composite_explicit_autograd_non_functional_kernel
        ) or (
            self.has_composite_implicit_autograd_kernel
            and self.has_composite_implicit_autograd_nested_tensor_kernel
        )
```

File: aten/src/ATen/native/nested/NestedTensorMath.cpp
Comment (albanD) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:719]: Always use `TORCH_CHECK*`

**Review by albanD — COMMENTED** [review]
SGTM
@drisspg I think we want to import this before landing it to see if internal changes are necessary. Can you take care of that?

**Comment by drisspg** [comment]
> SGTM @drisspg I think we want to import this before landing it to see if internal changes are necessary. Can you take care of that?

Will work through that now

**Comment by drisspg** [comment]
Broke your PR trying to rebase

File: test/test_nestedtensor.py
Comment (drisspg) [reply test/test_nestedtensor.py:238]: yeah because contiguous calls `clone` under the hood. And previously we did not support MemoryFormat::Contiguous but his PR introduced the change

**Comment by YifanShenSZ** [comment]
I'll leave it to you guys :stuck_out_tongue_winking_eye:

**Review by bdhirsh — APPROVED** [review]
nice work!

@drisspg should probably be the one to merge, since he's working on appeasing the internal build

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**
